Src/VimCore/CommonOperations.fs

```diff
@@ -2387,19 +2427,385 @@ type internal CommonOperations
+            let visualSelection =
+                VisualSelection.CreateInitial visualKind caretPoint tabStop selectionKind useVirtualSpace
+            visualSelection.GetPrimarySelectedSpan selectionKind
+
+        // Collect the command result and new selected span or any embedded
+        // visual span, if present.
+        let getResultingSpan oldSelectedSpan result =
+            match getSwitchToVisualKind result with
+            | Some visualKind ->
+                getInitialSelection visualKind
+            | None ->
+                match visualModeKind with
+                | Some modeKind ->
+                    getVisualSelectedSpan modeKind oldSelectedSpan
+                | None ->
+                    match getVisualSelection result with
+                    | Some selectedSpan ->
+                        selectedSpan
+                    | None ->
+                        x.PrimarySelectedSpan
+
+        // Set a temporary visual anchor point.
+        let setVisualAnchorPoint (anchorPoint: VirtualSnapshotPoint) =
+            if Option.isSome visualModeKind then
+                let snapshot = _textBuffer.CurrentSnapshot
+                let position = anchorPoint.Position.Position
+                let trackingPoint =
+                    snapshot.CreateTrackingPoint(position, PointTrackingMode.Negative)
+                    |> Some
+                _vimBufferData.VisualAnchorPoint <- trackingPoint
+
+        // Get the results for all actions.
+        let getResults () =
+            seq {
+                let indexedSpans =
+                    selectedSpans
+                    |> Seq.mapi (fun index span -> index, span)
+
+                // Iterate over all selections.
+                for index, oldSelectedSpan in indexedSpans do
+
+                    // Set the buffer local caret index.
+                    _vimBufferData.CaretIndex <- index
+
+                    // Set the visual anchor point.
+                    setVisualAnchorPoint oldSelectedSpan.AnchorPoint
+
+                    // Temporarily set the real caret and selection.
+                    x.MapSelectedSpanToCurrentSnapshot oldSelectedSpan
+                    |> x.SetTemporarySelectedSpan
+
+                    // Run the action once and get the result.
+                    let result = runActionAndCompleteTransaction action
+
+                    // Get the resulting span.
+                    let newSelectedSpan =
+                        getResultingSpan oldSelectedSpan result
+
+                    let newAnchorPoint =
+                        _vimBufferData.VisualAnchorPoint
+
+                    yield result, newSelectedSpan, newAnchorPoint
+            }
+            |> Seq.toList
+
+        // Do the action for all selections
+        let doActions () =
+
+            // Create a linked transaction for the overall operation.
+            let transaction = createTransaction()

```

Comment:
This should be a `use` instead of `let` to ensure the transaction is closed out in exceptional circumstances.

Reply:
No, this transaction can be left open, say, if we are switching to insert mode. But it is true we should complete the transaction if there were an exception.

Reply:
Without catching and re-raising any exception, I had to use a mutable flag to properly close the transaction. Not sure if there is a more elegant way of accomplishing the desired effect.

- No exceptions, command switches to insert -> use and leave open
- No exceptions, command does not switch to insert -> close
- Exception -> Don't interfere but close the transaction